providers/base/tests/test_pipewire_utils.py

```diff
@@ -0,0 +1,634 @@
+    def test_succ(self, mock_checkout):
+        pt = PipewireTest()
+
+        # detect audio sink succ
+        mock_checkout.return_value = self.sink_audio_node
+        self.assertEqual(PipewireTestError.NO_ERROR,
+                         pt.detect_device("audio", "sink"))
+
+        # detect audio source succ
+        mock_checkout.return_value = self.source_audio_node
+        self.assertEqual(PipewireTestError.NO_ERROR,
+                         pt.detect_device("audio", "source"))
+
+    @patch("subprocess.check_output")
+    def test_fail(self, mock_checkout):
+        pt = PipewireTest()
+
+        # detect audio sink fail
+        mock_checkout.return_value = self.source_audio_node
+        self.assertEqual(PipewireTestError.NOT_DETECTED,
+                         pt.detect_device("audio", "sink"))
+
+        # detect audio source fail
+        mock_checkout.return_value = self.sink_audio_node
+        self.assertEqual(PipewireTestError.NOT_DETECTED,
+                         pt.detect_device("audio", "source"))
+
+        # wrong media class
+        mock_checkout.return_value = self.sink_audio_node
+        self.assertEqual(PipewireTestError.NOT_DETECTED,
+                         pt.detect_device("xxx", "ooo"))
+
+
+class SelectDeviceTests(unittest.TestCase):
+
+    # Correct Node
+    sink_audio_node = '[{\
+                        "id": 29,\
+                        "type": "PipeWire:Interface:Node",\
+                        "version": 3,\
+                        "permissions": [ "r", "w", "x", "m" ],\
+                        "info": {\
+                          "max-input-ports": 0,\
+                          "max-output-ports": 0,\
+                          "change-mask": [ "input-ports",\
+                                           "output-ports",\
+                                           "state", "props",\
+                                           "params" ],\
+                          "n-input-ports": 0,\
+                          "n-output-ports": 0,\
+                          "state": "suspended",\
+                          "error": null,\
+                          "props": {\
+                            "media.class": "Audio/Sink",\
+                            "factory.name": "support.node.driver",\
+                            "node.name": "demo-hdmi",\
+                            "node.description": "unit test",\
+                            "priority.driver": 19000,\
+                            "node.group": "pipewire.freewheel",\
+                            "node.freewheel": true,\
+                            "factory.id": 10,\
+                            "clock.quantum-limit": 8192,\
+                            "node.driver": true,\
+                            "object.id": 29,\
+                            "object.serial": 29\
+                          },\
+                          "params": {\
+                          }\
+                        }\
+                      }]'
```

Comment:
please use multiline strings

Reply:
Change this to suggestion according to code review criteria, feel free to close it.